Create hermetic build environment

fortuna opened the pull request:
This introduces the `build_env` tool that let's you run a pre-built environment. It adds support for building new images and publishing to https://hub.docker.com/r/uproxy/build/tags/.

See the [README.md](https://github.com/uProxy/uproxy/blob/fortuna-build-docker/tools/build_env/README.md) for details on how to use it.

This PR pretty much deprecates #2479

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2487)
<!-- Reviewable:end -->

trevj commented on tools/build_env/Dockerfile:
```diff
@@ -0,0 +1,44 @@
+# See README.md file for how this image is used.
+
+FROM selenium/node-chrome
```
This is a large image - and I guess it doesn't include Firefox. Could this explain why some of the tests still fail?

Anyway, we can think about moving to a difference baseimage at some point in the future.

trevj commented on tools/build_env/build_env.sh:
```diff
@@ -0,0 +1,57 @@
+      -it $(get_image_version) "$@"
+}
+
+function main() {
```
Looks to me like you could probably eliminate this `main` function without sacrificing any readability :-)

trevj commented on tools/build_env/entrypoint.sh:
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+
+Xvfb :10 -screen 0 1280x1024x24 &
```
Oooooh really wish we could remove this...X is only needed for integration tests.

trevj commented on tools/build_env/README.md:
`````diff
@@ -0,0 +1,75 @@
+
+Use the `enter` command:
+````
+./tools/build_env/build_env.sh enter
`````
Perhaps enter should be the default option?

trevj commented on tools/build_env/README.md:
```diff
@@ -0,0 +1,75 @@
+
+This will:
+
+1. Start a new Docker container with the `uproxy/build:<tag>` image
```
You need a leading space on these items for the markdown renderer to make it a list (Github's preview feature is handy for tweaking this formatting stuff).

trevj commented on tools/build_env/build_env.sh:
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+set -eu
```
A pattern in our shell scripts has developed over in the uproxy-docker repo, you might consider using it: it uses `getopts` and offers help and could help here (I ran the script several times, wondering why I saw no output, before I read the README a little closer):
https://github.com/uProxy/uproxy-docker/blob/master/testing/run-scripts/run_cloud.sh#L25

trevj commented:
Nice.

I see how the mix of inputs and outputs in `build/` is really missing up here - we're bending over backwards here to accomodate it and I'm sceptical we'll be diligent in keeping the images very much up to date :-) That said, I do like where this is heading! I had a few comments but then I think we'll be good to :+1:.

fortuna commented:
Review status: 0 of 6 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[tools/build_env/build_env.sh, line 2 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP4-r1-2:-KMBq0KfkShaPjN6z2ti:1828283163) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/build_env.sh#L2)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> A pattern in our shell scripts has developed over in the uproxy-docker repo, you might consider using it: it uses `getopts` and offers help and could help here (I ran the script several times, wondering why I saw no output, before I read the README a little closer):
> 
> https://github.com/uProxy/uproxy-docker/blob/master/testing/run-scripts/run_cloud.sh#L25
</details>
I added a help message. Notice that now if you don't specify any command you will enter the build environment.

I think the example you sent is a somewhat different use case. `getopts` is useful for specifying and parsing flags, but my script has no flags.

---

*[tools/build_env/build_env.sh, line 52 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP4-r1-52:-KMAESLKyUTLJyjPZbWU:631437939) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/build_env.sh#L52)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Looks to me like you could probably eliminate this `main` function without sacrificing any readability :-)
</details>
I guess this is mostly a personal style preference. I use `main` for a few reasons:

*  Some bash features are only allowed inside functions:
  * `return` calls (which I use here)
  * Definition of local variables
* For readability: `main` makes it clearer that there isn't any code outside the function, so I don't need to scan the entire file for code between the function definition.

---

*[tools/build_env/Dockerfile, line 3 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP1-r1-3:-KMBBl5vQ_rGeA_4hI2D:65624307) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/Dockerfile#L3)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> This is a large image - and I guess it doesn't include Firefox. Could this explain why some of the tests still fail?
> 
> 
> 
> Anyway, we can think about moving to a difference baseimage at some point in the future.
</details>
#### Image Size
The image was huge indeed. After a number of optimizations it went down from 679MB to 432MB.
Some optimizations:

* Revert to `ubuntu:trusty` as the base image and explicitly installing Google Chrome.
* Add a number of cleanup commands to `RUN` calls. I learned that I can't just do a big cleanup at the end because the image has all the intermediate states :-( (see [docker-squash](http://jasonwilder.com/blog/2014/08/19/squashing-docker-images/)).

#### Test Failure
I couldn't figure out what the problem is. The error I get on every `jasmine:` target is

    Warning: Cannot read property 'pid' of undefined Used --force, continuing.

I don't think Firefox is the problem. AFAICT, we don't use real browsers for those tests, and they fail even when I install Firefox. 

---

*[tools/build_env/entrypoint.sh, line 3 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP5-r1-3:-KMBBJFFsNpLPLo8CFwr:-1337331610) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/entrypoint.sh#L3)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Oooooh really wish we could remove this...X is only needed for integration tests.
</details>
We can. We just need to move that set up to the test task that needs it.
Since I don't want to mess with the grunt files in this change (and I don't yet properly understand them), I've added a TODO

---

*[tools/build_env/README.md, line 28 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP3-r1-28:-KMALr4BcM00XdmiRxXu:1367071830) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/README.md#L28)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> Perhaps enter should be the default option?
</details>
Good suggestion. Done

---

*[tools/build_env/README.md, line 33 \[r1\]](https://reviewable.io:443/reviews/uproxy/uproxy/2487#-KKVQVmr4r-3YjoJnLP3-r1-33:-KMAMDihyjCTo0pemTJz:-1215427487) ([raw file](https://github.com/uproxy/uproxy/blob/333ab7c5f8d19a805f1ab27fd5b1200ac09c961c/tools/build_env/README.md#L33)):*
<details><summary><i>Previously, trevj wrote…</i></summary>
> You need a leading space on these items for the markdown renderer to make it a list (Github's preview feature is handy for tweaking this formatting stuff).
</details>
No need for space. Github and Atom's markdown preview  are both showing the list properly:
https://screenshot.googleplex.com/heM5kiPqUhJ

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/uproxy/uproxy/2487)*
<!-- Sent from Reviewable.io -->
